curvefs/src/metaserver/trash.cpp

```diff
@@ -52,27 +54,38 @@ void TrashImpl::Init(const TrashOption &option) {
     options_ = option;
     s3Adaptor_ = option.s3Adaptor;
     mdsClient_ = option.mdsClient;
+    copysetNode_ = copyset::CopysetNodeManager::GetInstance().
+        GetSharedCopysetNode(poolId_, copysetId_);
     isStop_ = false;
 }
 
-void TrashImpl::Add(uint32_t fsId, uint64_t inodeId, uint32_t dtime) {
-    TrashItem item;
-    item.fsId = fsId;
-    item.inodeId = inodeId;
-    item.dtime = dtime;
+void TrashImpl::StopScan() {
+    isStop_ = true;
+}
 
-    LockGuard lg(itemsMutex_);
+bool TrashImpl::IsStop() {
+    return isStop_;
+}
+
+void TrashImpl::Add(uint64_t inodeId, uint64_t dtime) {
     if (isStop_) {
         return;
     }
-    trashItems_.push_back(item);
-    VLOG(6) << "Add Trash Item success, item.fsId = " << item.fsId
-            << ", item.inodeId = " << item.inodeId
-            << ", item.dtime = " << item.dtime;
+    LockGuard lg(itemsMutex_);
+    trashItems_.emplace_back(TrashItem(inodeId, dtime));
+    VLOG(6) << "Add Trash Item success, fsId = " << fsId_
+            << ", partitionId = " << partitionId_
+            << ", inodeId = " << inodeId
+            << ", dtime = " << dtime;
 }
 
 void TrashImpl::ScanTrash() {
     LockGuard lgScan(scanMutex_);
+    // only scan on leader
+    if (copysetNode_ == nullptr || !copysetNode_->IsLeaderTerm()) {
```

Comment:
The following cycle also needs to be interrupted when the node is not the leader.